### PR TITLE
Support sitemaps created through UI in proxy

### LINKED ISSUE #2154: Video in sitemap exception
Same issue as the Image Renderer, so I am going to just defer to that issue : https://github.com/openhab/openhab-webui/issues/763

Tested on SNAPSHOT build : 
runtimeInfo:
  version: 3.1.0
  buildString: "Build #2163"
locale: en_US
systemInfo:
  configFolder: /etc/openhab
  userdataFolder: /var/lib/openhab
  logFolder: /var/log/openhab
  javaVersion: 11.0.10
  javaVendor: Azul Systems, Inc.
  javaVendorVersion: Zulu11.45+27-CA
  osName: Linux
  osVersion: 5.4.0-56-generic
  osArchitecture: amd64
  availableProcessors: 64
  freeMemory: 2104767072
  totalMemory: 2592079872
bindings:
  - amazonechocontrol
  - astro
  - chromecast
  - harmonyhub
  - icalendar
  - mqtt
  - plex
  - pushover
  - remoteopenhab
  - sonos
  - speedtest
  - zoneminder
  - zwave
clientInfo:
  device:
    ios: false
    android: false
    androidChrome: false
    desktop: true
    iphone: false
    ipod: false
    ipad: false
    edge: false
    ie: false
    firefox: false
    macos: false
    windows: true
    cordova: false
    phonegap: false
    electron: false
    nwjs: false
    webView: false
    webview: false
    standalone: false
    os: windows
    pixelRatio: 1
    prefersColorScheme: light
  isSecureContext: false
  locationbarVisible: true
  menubarVisible: true
  navigator:
    cookieEnabled: true
    deviceMemory: N/A
    hardwareConcurrency: 8
    language: en-US
    languages:
      - en-US
      - en
    onLine: true
    platform: Win32
  screen:
    width: 1920
    height: 1080
    colorDepth: 24
  support:
    touch: false
    pointerEvents: true
    observer: true
    passiveListener: true
    gestures: false
    intersectionObserver: true
  themeOptions:
    dark: light
    filled: true
    pageTransitionAnimation: default
    bars: filled
    homeNavbar: default
    homeBackground: default
    expandableCardAnimation: default
  userAgent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML,
    like Gecko) Chrome/87.0.4280.141 Safari/537.36
timestamp: 2021-01-25T05:24:33.582Z

Same failure for the most part as the ImageRenderer, but the VideoRenderer doesn't have a check for sitemap being null so it just throws an exception : 
`00:14:55.120 [tp1751055337-66] WARN  org.eclipse.jetty.server.HttpChannel:600 - /basicui/app
java.lang.NullPointerException: null
	at org.openhab.ui.basic.internal.render.VideoRenderer.renderWidget(VideoRenderer.java:69)
	at org.openhab.ui.basic.internal.render.PageRenderer.renderWidget(PageRenderer.java:195)
	at org.openhab.ui.basic.internal.render.PageRenderer.processChildren(PageRenderer.java:161)
	at org.openhab.ui.basic.internal.render.PageRenderer.processPage(PageRenderer.java:125)
	at org.openhab.ui.basic.internal.servlet.WebAppServlet.service(WebAppServlet.java:183)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:725)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:852)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:544)
	at org.ops4j.pax.web.service.jetty.internal.HttpServiceServletHandler.doHandle(HttpServiceServletHandler.java:71)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:536)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1581)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1307)
	at org.ops4j.pax.web.service.jetty.internal.HttpServiceContext.doHandle(HttpServiceContext.java:298)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:482)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1549)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1204)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.ops4j.pax.web.service.jetty.internal.JettyServerHandlerCollection.handle(JettyServerHandlerCollection.java:82)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.Server.handle(Server.java:494)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:374)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:268)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produce(EatWhatYouKill.java:135)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:782)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:918)
	at java.base/java.lang.Thread.run(Thread.java:834)`

Exception @ : 
`String widgetId = itemUIRegistry.getWidgetId(w);
        String sitemap = w.eResource().getURI().path();`

To stop it from throwing an exception should add a check on sitemap like it's done in the ImageRenderer : 
`        String sitemap = null;
        if (w.eResource() != null) {
            sitemap = w.eResource().getURI().path();
        }`

Though it would still display nothing until the underlying issue is fixed(would seem to be core related).

### PR DESCRIPTION
**Pull request by bigbasec**
Third attempt at this PR, but this very simply allows for the proxy to find the sitemaps for both file based as well as gui generated.

Fixes #2154
Fixes openhab/openhab-webui#763
Fixes openhab/openhab-webui#745

### CODE REVIEW

**Comment by kaikreuzer** [comment]
@openhab/core-maintainers As I have heavily contributed to this PR in a second commit, I'd like to ask one of you to do the review&merge. Thanks!

File: bundles/org.openhab.core.ui/src/main/java/org/openhab/core/ui/internal/proxy/ProxyServletService.java
Comment (cweitkamp) [inline bundles/org.openhab.core.ui/src/main/java/org/openhab/core/ui/internal/proxy/ProxyServletService.java:101]: Should we use a `CopyOnWriteArrayList` here to avoid concurrency issues?

**Review by cweitkamp — COMMENTED** [review]
Thank you both. I left one question.

Can you please link the related issues again? For tracking and documentation.

File: bundles/org.openhab.core.ui/src/main/java/org/openhab/core/ui/internal/proxy/ProxyServletService.java
Comment (bigbasec) [reply bundles/org.openhab.core.ui/src/main/java/org/openhab/core/ui/internal/proxy/ProxyServletService.java:101]: I'll defer to @kaikreuzer in regards to the concurrency possibility as he would know the upstream way better than myself.   

The issues this covers : 
https://github.com/openhab/openhab-webui/issues/763
https://github.com/openhab/openhab-core/issues/2154
https://github.com/openhab/openhab-webui/issues/745

This is a 2 part fix, the second part if here I think awaiting the merge of this one : 
https://github.com/openhab/openhab-webui/pull/860

Comment (kaikreuzer) [reply bundles/org.openhab.core.ui/src/main/java/org/openhab/core/ui/internal/proxy/ProxyServletService.java]: I've addressed the comment and also added the fixed issues to the PR description.

**Review by cweitkamp — APPROVED** [review]

**Comment by t2000** [comment]
Could it be possible that you broke the feature with this PR?
I am using a sitemap file since ages and it was working fine in OH3 until I recently updated to a new snapshot: https://ci.openhab.org/job/openHAB3-Distribution/2196/ which should include your fix.

My sitemap is no showing the camera picture anymore:

It is defined (and never changed, thus it worked before) as:
```
Frame label="Kamera" {
		Text label="Haustür" icon="camera"{
			Image label="Haustür" icon="camera" url="http://doorbell:8080/picture.jpg" refresh=500			
		}
	}
```

I also tried to replace the url with the IP to rule out DNS issues and also tried to use an external https url:   https://de.wikipedia.org/wiki/Testbild#/media/Datei:SW_Testbild_auf_Philips_TD1410U.jpg but it doesnt work.

sitemap-no-picture-dev

**Comment by bigbasec** [comment]
Could also be that the basic UI changes weren't merged yet.   This was a 2 part fix, one to fix the proxy and one to fix the basic UI.   I'll take a look and see what's going on here.

**Comment by bigbasec** [comment]
Yea, this is going to require https://github.com/openhab/openhab-webui/pull/860 to be merged to actually fix the problem.   

Issue is everywhere in the UI's(basic too) the sitemaps are recognized by their real name "MySitemap", the ".sitemap" is added by the core when it loads the files.   That's why when you fire up Basic UI and look at the location it shows (http://localhost:8080/basicui/app?sitemap=TestFileBased), the .sitemap is not there.   The change I made there is basically making them both common, so we don't need 2 different handlers depending on how the sitemap was made.   

So the problem here is for Images and Videos specifically(because they use the proxy) I submitted the change to the core to fix the issue where Images and Video would not work on anything UI generated.   It now does a lookup on all providers(both UI and file generated) by name, the way the UI and BasicUI work.  

As part of that I submitted one for the UI to make the URL's for both UI generated and file based the same, regardless of how they were configured on the system.    Example of how it is now : 
UI Generated : 
../proxy?sitemap=UIGeneratedSitemap&amp;widgetId=00
File generated : 
../proxy?sitemap=FileGeneratedSitemap.sitemap&amp;widgetId=00

Problem is the core is now going to look those up and fail on the one w/ .sitemap added to it(as that's it's filename).

**Comment by t2000** [comment]
@bigbasec Thanks for the detailed explanation. As the other PR is also merged, I will give it a test in the next couple of days.

**Comment by t2000** [comment]
@bigbasec Just FYI: I just tested a new snapshot that includes https://github.com/openhab/openhab-webui/pull/860 and my camera picture is now correctly shown in my sitemap again. Thanks for fixing it!

**Comment by t2000** [comment]
@bigbasec Unfortunately I have to report that basicUI is now working, BUT the android app is not :(

I am using my sitemap to be shown in the android app, but I do not get a picture :(

Is there anything missing on the android app side? shall I create an issue on the android project?

**Comment by bigbasec** [comment]
> @bigbasec Unfortunately I have to report that basicUI is now working, BUT the android app is not :(
> 
> I am using my sitemap to be shown in the android app, but I do not get a picture :(
> 
> Is there anything missing on the android app side? shall I create an issue on the android project?

Yup, it's because the rest interface is appending a ".sitemap" on the proxy URL.   I'll submit a PR for this in a little bit that'll take care of that issue.   Just want to test it out before I do.   Thanks for the report!

**Pull request merged**
